pkg/operator/edgefs/isgw/isgw.go

```diff
@@ -0,0 +1,417 @@
+	return c.CreateOrUpdate(s, true, ownerRefs)
+}
+
+// Start the isgw instance
+func (c *ISGWController) CreateOrUpdate(s edgefsv1alpha1.ISGW, update bool, ownerRefs []metav1.OwnerReference) error {
+	logger.Infof("starting update=%v service=%s", update, s.Name)
+
+	logger.Infof("ISGW Base image is %s", c.rookImage)
+	// validate ISGW service settings
+	if err := validateService(c.context, s); err != nil {
+		return fmt.Errorf("invalid ISGW service %s arguments. %+v", s.Name, err)
+	}
+
+	if s.Spec.ReplicationType == "" {
+		s.Spec.ReplicationType = defaultReplicationType
+	}
+
+	if s.Spec.DynamicFetchPort == 0 {
+		s.Spec.DynamicFetchPort = defaultDynamicFetchPort
+	}
+
+	if s.Spec.LocalAddr == "" {
+		s.Spec.LocalAddr = defaultLocalIPAddr + ":" + strconv.Itoa(defaultLocalPort)
+	}
+
+	// check if ISGW service already exists
+	exists, err := serviceExists(c.context, s)
+	if err == nil && exists {
+		if !update {
+			logger.Infof("ISGW service %s exists in namespace %s", s.Name, s.Namespace)
+			return nil
+		}
+		logger.Infof("ISGW service %s exists in namespace %s. checking for updates", s.Name, s.Namespace)
+	}
+
+	// start the deployment
+	deployment := c.makeDeployment(s.Name, s.Namespace, c.rookImage, s.Spec)
+	if _, err := c.context.Clientset.ExtensionsV1beta1().Deployments(s.Namespace).Create(deployment); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create %s deployment. %+v", appName, err)
+		}
+		logger.Infof("%s deployment already exists", appName)
+	} else {
+		logger.Infof("%s deployment started", appName)
+	}
+
+	// create the isgw service
+	service := c.makeISGWService(instanceName(s.Name), s.Name, s.Namespace, s.Spec)
+	if _, err := c.context.Clientset.CoreV1().Services(s.Namespace).Create(service); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create ISGW service. %+v", err)
+		}
+		logger.Infof("ISGW service %s already exists", service)
+	} else {
+		logger.Infof("ISGW service %s started", service)
+	}
+
+	return nil
+}
+
+func getDirection(isgwSpec edgefsv1alpha1.ISGWSpec) int {
+	direction := 0
+	if isgwSpec.Direction == "send" {
+		direction = 1
+	} else if isgwSpec.Direction == "receive" {
+		direction = 2
+	} else {
+		direction = 3
+	}
+	return direction
+}
+
+func (c *ISGWController) makeISGWService(name, svcname, namespace string, isgwSpec edgefsv1alpha1.ISGWSpec) *v1.Service {
+	direction := getDirection(isgwSpec)
+	labels := getLabels(name, svcname, namespace)
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: v1.ServiceSpec{
+			Selector: labels,
+			Type:     v1.ServiceTypeClusterIP,
+			Ports: []v1.ServicePort{
+				{Name: "grpc", Port: 49000, Protocol: v1.ProtocolTCP},
+			},
+		},
+	}
+
+	if direction == 2 || direction == 3 {
+		larr := strings.Split(isgwSpec.LocalAddr, ":")
```

Comment:
Please use the net.IP.ParseIP() function (see https://golang.org/pkg/net/#ParseIP).
If someone would uses an IPv6 addresses this would fail.

Reply:
love net library. Fixed!!!